### PR TITLE
 storageccl: scan all versions of changed key in MVCCIncrementalIterator 

### PR DESCRIPTION
**Pull request by pbardea**
This PR re-introduces #45163. It needs to be merged after #45759.

This latest implementation of MVCCIncrementalIterator updates its use of
the time-bound iterator to only be used to skip over keys that were not
modified at all within the time bound limits.

Previously, the incremental iterator primarily used the time-bound
iterator and then used a non-time-bound iterator (sanity_iter) to check
that meta keys were properly read. See #43799 for discussion around why
this may not be sufficient.

This commit updates the implementation to use the time-bound iterator
more conservatively. The 2 iterators that it now uses are `iter` (a
non-time-bound iterator and `timeBoundIter`). The main iteration is now
done with `iter`. When `iter` sees a new key, the `timeBoundIter` is
incremented and if it goes further than `iter`, `iter` is `Seek`ed
forward to the meta key of the new key. This means that `iter` should
visit every version of every key that `timeBoundIter` visits.

It should be noted that if many adjacent keys were modified, the above
logic tries to avoid to do a `Seek` to the next key as a `Seek` is ~an
order of magnitude more expensive than a `Next`.

In cases where a lot of keys were modified, we expect to do ~double the
work since 2 iterators would need to seek to every key. However, the
expected use case for the MVCCIncrementalIterator is such that the
_vast_ majority of the data is excluded from the range (as should be the
case for incremental backups), since there is potentially years worth of
data that can be skipped over, versus hours worth of data to include.

Release note: None

### CODE REVIEW

**Comment by pbardea** [comment]
Note that this PR differs slightly from the first implementation:
```
diff --git a/c-deps/libroach/incremental_iterator.cc b/c-deps/libroach/incremental_iterator.cc
index 837f2bffa8..b613dfbad3 100644
--- a/c-deps/libroach/incremental_iterator.cc
+++ b/c-deps/libroach/incremental_iterator.cc
@@ -79,7 +79,7 @@ void DBIncrementalIterator::maybeSkipKeys() {
     return;
   }
   rocksdb::Slice iter_key;
-  if(!extractKey(time_bound_iter->rep->key(), &iter_key)) {
+  if(!extractKey(iter_key->rep->key(), &iter_key)) {
     return;
   }
```

This should be followed up with another test that ensures that we're actually skipping some keys with this optimization.

File: c-deps/libroach/incremental_iterator.cc [hunk 1]
```diff
@@ -95,20 +51,106 @@ bool DBIncrementalIterator::legacyTimestampIsLess(const cockroach::util::hlc::Le
...
+    }
+
+    auto cmp = iter_key.compare(tbi_key);
+    // The case where iterKey.Compare(tbiKey) == 0 is the fast-path is when the
+    // TBI and the main iterator are in lockstep. In this case, the main
+    // iterator was referencing the next key that would be visited by the TBI.
+    // This means that for the incremental iterator to perform a Next or NextKey
+    // will require only 1 extra NextKey invocation while they remain in
+    // lockstep.
+
```
Comment (miretskiy) [inline c-deps/libroach/incremental_iterator.cc]: Do you think this comment belongs outside this if statement?
I mean it talks about fastpath (iter_key == tbi_key).  

Also, I think it's easier to read if you use e.g. "iter_key == tbi_key" in the comment as opposed to specifics (Compare)

File: c-deps/libroach/incremental_iterator.cc [hunk 2]
```diff
@@ -95,20 +51,106 @@ bool DBIncrementalIterator::legacyTimestampIsLess(const cockroach::util::hlc::Le
...
+    if (cmp > 0) {
+      // If the tbiKey is still behind the iterKey, the TBI key may be seeing
+      // phantom MVCCKey.Keys. These keys may not be seen by the main iterator
+      // due to aborted transactions and keys which have been subsumed due to
+      // range tombstones. In this case we can SeekGE() the TBI to the main iterator.
+
+      // NB: Seek() is expensive, and this is only acceptable since we expect
+      // this case to rarely occur.
```
Comment (miretskiy) [inline c-deps/libroach/incremental_iterator.cc]: I feel like this important note belongs right at the beginning of this if statement -- it explains why we call Next, as opposed to seeking right away.

**Review by miretskiy — APPROVED** [review]
I'm having problems submitting my pr due to flaky test -- so make sure you wait for that pr to hit the master.

**Review by sumeerbhola — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45785)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @pbardea, and @sumeerbhola)

___
*[c-deps/libroach/incremental_iterator.cc, line 105 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45785#-M1ipOQyPKT-fWjmUZJf-r2-105:-M1kQVsA1TeIFiJzbQZ4:b-sojpdd) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ad8ccfc283013f7211557680eb3fa7ec6bcc7f1/c-deps/libroach/incremental_iterator.cc#L105)):*
> Do you think this comment belongs outside this if statement?
> I mean it talks about fastpath (iter_key == tbi_key).

I think there are three cases (if we have any incremental backup metrics we should keep a counter for each. Or maybe just log?):
- tbi was one behind the iter: The next call on tbi will make them equal. This is the fast path mentioned here. I think this comment could be elevated just before the call to `DBIterNext(time_bound_iter.get(), true /* skip_current_key_versions */);` since it motivates calling next instead of seek. This could be common if most keys are modified or the modifications are clustered in the key space.
- the next call on tbi moves it far ahead of iter. This will be the common case when modifications are sparse in the key space. This may turn out to be the real common case -- in which case the fast path did not help.
- the next call on tbi keeps it behind the iter (should be very rare).

**Review by pbardea — COMMENTED** [review]
Update on this issue: I've moved the comments around based on the comments, but I am still seeing the libroach assertion failing on this branch. Investigating further.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45785#-:-M1m-KqR1f8xm1QukFDu:bhht66)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)

___
*[c-deps/libroach/incremental_iterator.cc, line 113 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45785#-M1ipOQyPKT-fWjmUZJf-r2-113:-M1l1Ihr6fehwedyR2SE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ad8ccfc283013f7211557680eb3fa7ec6bcc7f1/c-deps/libroach/incremental_iterator.cc#L113)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I feel like this important note belongs right at the beginning of this if statement -- it explains why we call Next, as opposed to seeking right away.
</blockquote></details>

Done.

**Review by miretskiy — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45785)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @pbardea)

___
*[c-deps/libroach/incremental_iterator.h, line 22 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45785#-M1zTk8N10YEMLYzwNd9:-M1zTk8N10YEMLYzwNdA:b-ribyq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/37fd6d58e4175de2b221c99bf95ea2599b8773ed/c-deps/libroach/incremental_iterator.h#L22)):*
> ```
> which is used as an oracle to test DBIncrementalIterator.
> ```
Is it going to be used only as an oracle? Or, eventually, it will be the iterator to be used w/ pebble?
___
*[c-deps/libroach/incremental_iterator.cc, line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45785#-M1zU5ErC6SzgJ4aZcR-:-M1zU5ErC6SzgJ4aZcR0:b-89qalx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d117fcec132c9a8b20e98e70ec1f1673f14fb775/c-deps/libroach/incremental_iterator.cc#L49)):*
> ```cpp
> // boolean indicating whether t1 is less than t2.
> bool DBIncrementalIterator::legacyTimestampIsLess(const cockroach::util::hlc::LegacyTimestamp& t1,
>                                                   const cockroach::util::hlc::LegacyTimestamp& t2) {
> ```

might wrapping these long lines?  c++ usually tries to stay under 80 cols
___
*[c-deps/libroach/incremental_iterator.cc, line 56 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45785#-M1zUAf37XGwq5gqpDPY:-M1zUAf37XGwq5gqpDPZ:b-7h39o0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d117fcec132c9a8b20e98e70ec1f1673f14fb775/c-deps/libroach/incremental_iterator.cc#L56)):*
> ```cpp
> // extractKey extracts the key portion of the mvcc_key and put it in key. It
> // returns a validity indicator.
> WARN_UNUSED_RESULT bool DBIncrementalIterator::extractKey(rocksdb::Slice mvcc_key, rocksdb::Slice *key) {
> ```

wrap?
___
*[c-deps/libroach/incremental_iterator.cc, line 250 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45785#-M1zWNby0bEVuOITgkBp:-M1zWNby0bEVuOITgkBq:b-jy47dy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d117fcec132c9a8b20e98e70ec1f1673f14fb775/c-deps/libroach/incremental_iterator.cc#L250)):*
> ```cpp
>     }
>     const rocksdb::Slice tbi_key(time_bound_iter->rep->key());
>     const rocksdb::Slice iter_key(ToSlice(key.key));
> ```

This line deserves an "NB: comment" ...

**Review by pbardea — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45785)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)

___
*[c-deps/libroach/incremental_iterator.h, line 22 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45785#-M1zTk8N10YEMLYzwNd9:-M1zeb91CH2sEcHOZr4P:bkqrw5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/37fd6d58e4175de2b221c99bf95ea2599b8773ed/c-deps/libroach/incremental_iterator.h#L22)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
> which is used as an oracle to test DBIncrementalIterator.
> ```
Is it going to be used only as an oracle? Or, eventually, it will be the iterator to be used w/ pebble?
</blockquote></details>

It seems reasonable that it'll be used as the iterator when we move over to Pebble. I added a note which mentions that this iterator may be useful when Pebble becomes the default store.
___
*[c-deps/libroach/incremental_iterator.cc, line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45785#-M1zU5ErC6SzgJ4aZcR-:-M1zlld1AkPpkhi_haqW:bt26rmj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d117fcec132c9a8b20e98e70ec1f1673f14fb775/c-deps/libroach/incremental_iterator.cc#L49)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

might wrapping these long lines?  c++ usually tries to stay under 80 cols
</blockquote></details>

Done. I ran the formatter (https://github.com/cockroachdb/cockroach/blob/master/c-deps/libroach/README.md). I saw that there were formatting changes to be made to other files that I didn't touch - it may be worth doing another PR with those changes.
___
*[c-deps/libroach/incremental_iterator.cc, line 56 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45785#-M1zUAf37XGwq5gqpDPY:-M1zm59p35yxkWQjDiba:b-cxfbnl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d117fcec132c9a8b20e98e70ec1f1673f14fb775/c-deps/libroach/incremental_iterator.cc#L56)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

wrap?
</blockquote></details>

Same as previous comment.
___
*[c-deps/libroach/incremental_iterator.cc, line 250 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45785#-M1zWNby0bEVuOITgkBp:-M1znHyt8VZU4jqu8uxC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d117fcec132c9a8b20e98e70ec1f1673f14fb775/c-deps/libroach/incremental_iterator.cc#L250)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This line deserves an "NB: comment" ... 
</blockquote></details>

Done.

**Review by pbardea — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45785)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)

___
*[c-deps/libroach/incremental_iterator.cc, line 105 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45785#-M1ipOQyPKT-fWjmUZJf-r2-105:-M1zznpY69jN1UFp2z-4:b1shbdp) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ad8ccfc283013f7211557680eb3fa7ec6bcc7f1/c-deps/libroach/incremental_iterator.cc#L105)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

> Do you think this comment belongs outside this if statement?
> I mean it talks about fastpath (iter_key == tbi_key).

I think there are three cases (if we have any incremental backup metrics we should keep a counter for each. Or maybe just log?):
- tbi was one behind the iter: The next call on tbi will make them equal. This is the fast path mentioned here. I think this comment could be elevated just before the call to `DBIterNext(time_bound_iter.get(), true /* skip_current_key_versions */);` since it motivates calling next instead of seek. This could be common if most keys are modified or the modifications are clustered in the key space.
- the next call on tbi moves it far ahead of iter. This will be the common case when modifications are sparse in the key space. This may turn out to be the real common case -- in which case the fast path did not help.
- the next call on tbi keeps it behind the iter (should be very rare).
</blockquote></details>

Updated the comments based on these suggestions.

**Review by miretskiy — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45785)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)

**Pull request merged**